spawn_worker/src/unix/spawn.rs

```diff
@@ -0,0 +1,266 @@
+            stdin: None,
+            stdout: None,
+            stderr: None,
+            target: Target::Noop,
+            inherit_env: true,
+            spawn_method: Default::default(),
+        }
+    }
+
+    pub fn target(&mut self, target: Target) -> &mut Self {
+        self.target = target;
+        self
+    }
+
+    pub fn stdin<T: Into<OwnedFd>>(&mut self, fd: T) -> &mut Self {
+        self.stdin = Some(fd.into());
+        self
+    }
+
+    pub fn stdout<T: Into<OwnedFd>>(&mut self, fd: T) -> &mut Self {
+        self.stdout = Some(fd.into());
+        self
+    }
+
+    pub fn stderr<T: Into<OwnedFd>>(&mut self, fd: T) -> &mut Self {
+        self.stderr = Some(fd.into());
+        self
+    }
+
+    pub fn spawn(&mut self) -> anyhow::Result<Child> {
+        let pid = self.do_spawn()?;
+
+        Ok(Child {
+            pid,
+            stdin: self.stdin.take(),
+            stderr: self.stderr.take(),
+            stdout: self.stdout.take(),
+        })
+    }
+
+    fn do_spawn(&self) -> anyhow::Result<Option<libc::pid_t>> {
+        let mut argv = ExecVec::default();
+        // set prog name (argv[0])
+        argv.push(CString::new("trampoline")?);
+        type SpawnFn = dyn Fn(&SealedExecVec, &SealedExecVec);
+
+        let spawn: Box<SpawnFn> = match &self.spawn_method {
+            #[cfg(target_os = "linux")]
+            SpawnMethod::FdExec => {
+                let fd = linux::write_trampoline()?;
+                Box::new(move |argv, envp| {
+                    // not using nix crate here, as it would allocate args after fork, which will lead to crashes on systems
+                    // where allocator is not fork+thread safe
+                    unsafe { libc::fexecve(fd.as_raw_fd(), argv.as_ptr(), envp.as_ptr()) };
+                    // if we're here then exec has failed
+                    panic!("{}", std::io::Error::last_os_error());
+                })
+            }
+            SpawnMethod::Exec => {
+                let path = CString::new(
+                    write_trampoline()?
+                        .into_temp_path()
+                        .keep()? // ensure the file is not auto cleaned in parent process
+                        .as_os_str()
+                        .to_str()
+                        .ok_or_else(|| anyhow::format_err!("can't convert tmp file path"))?,
+                )?;
+
+                Box::new(move |argv, envp| {
+                    // not using nix crate here, to avoid allocations post fork
+                    unsafe { libc::execve(path.as_ptr(), argv.as_ptr(), envp.as_ptr()) };
+                    // if we're here then exec has failed
+                    panic!("{}", std::io::Error::last_os_error());
+                })
+            }
+        };
+
+        match &self.target {
+            Target::Trampoline(f) => {
+                let (library_path, symbol_name) =
+                    match unsafe { crate::get_dl_path_raw(*f as *const libc::c_void) } {
```

Comment:
I think this looks good, though we should be extra careful as this is a place a bad user would want to use to inject his library.